### ISSUE TITLE
difference between valid "acc" and inference "WER"

### ISSUE THREAD

**Issue by LeeYongHyeok** [pytorch/fairseq#1448]
Hi, everyone.

Recently I ran the ASR example code. I was wondering why the difference between valid acc on train and WER of inference is so different.

In this case, valid set and test set is perfectly same. and I set the prameter, `MAX_TOKEN=5000` `N_BEST=1` & `BEAM=30` to use `infer.py`

![image](https://user-images.githubusercontent.com/39906282/69953569-dbea2e00-153c-11ea-88ec-fa23cbfb2792.png)

You can see the `valid acc` on test set, almost models are above `94%` but I got WER is about `17% ~ 27%`.

I know that `1-WER` and `acc` are different. But I don't understand the huge difference.

Additonaly, I found the strange decode problem. 

![image](https://user-images.githubusercontent.com/39906282/69962456-6177d900-1551-11ea-964e-013f98c37456.png)

Some sections of this sentence seem to have been reversed.

I wonder if others get similar results or if it's an infer code problem.

Thank you for reading this issue.

**Comment by lematt1991** [comment]
CC @alexeib

**Comment by liangzhenlin123** [comment]
@LeeYongHyeok Hi, I am also bothered for this issue, do you get any explanations? I want to reproduce the result in paper for examples/speech_recognition, however, I only achieve WER for 12.1 with a base transformer structure for Librispeech, and achieve about 11 in Aishell1, it is really wired... since the valid acc in Tensorboard almost get 95.0%...
Thanks a lot for u response ；D

**Comment by alexeib** [comment]
what "asr example code" did you run?

**Comment by liangzhenlin123** [comment]
@alexeib Just the demo which provided by fairseq. The path is fairseq/examples/speech_recognition, I have tried both vggtransformer_1 and vggtransformer_base in Librispeech and Aishell, however I cannot never reach a WER as 4.8 or 5.* as original paper said....

**Comment by alexeib** [comment]
@okhonko might be a better person to help with this

**Comment by LeeYongHyeok** [comment]
@liangzhenlin123 I used to modify the model of the example code. The above problem was caused by my mistake missing the VGGblocks for positional encoding of the vggtransformer model.

**Comment by liangzhenlin123** [comment]
@LeeYongHyeok Thanks for your reply! I directly run the demo's command line provided originally in README, so I think I don't face this mistake. So can you get the WER as well as this paper "Transformers with convolutiona lcontext for ASR"? BTW, whats your batchsize, have you used "--update-freq" to expand the max tokens?
Thanks a lot!

**Comment by LeeYongHyeok** [comment]
@liangzhenlin123 I can't remember the detailed parameters, quite a long time ago, but since I ran it with one GPU, maybe i set the `max token` to 8000 without fixed batch size. If you want a large batch size, I recommend using the `--fp16` option.

**Comment by liangzhenlin123** [comment]
@LeeYongHyeok Thanks a lot for your patience :).
